Explaining tagging (#266)
merely-useful/merely-useful.github.io: rse-git-advanced.Rmd

picking up others' changes and submitting their own whenever they want.
More importantly,
everyone has a chance to review work before it lands in the main repository.
As we discuss in Section \@ref(rse-style-review),
doing reviews doesn't just prevent errors from creeping in:
it is also an effective way to spread understanding and skills.

## How can I label a particular version of my project? {#rse-git-advanced-tag}

A [tag][tag] is a permanent label on a particular state of the repository.
Tags are theoretically redundant,
since the [full identifier][full-identifier-git] identifies that state as well,
but full identifiers are hard to remember and type,
and their [shortened versions][short-identifier-git] aren't any more meaningful.

Projects therefore use annotated tags to mark every major event in the project's history.
These tags are called "annotated" because
they allow their creators to add explanatory messages
just like [commit messages][commit-message].

> Research projects often use `report-date-event` for tag names,

Author: Changed.

such as `jse-2018-06-23-response` or `pediatrics-2018-08-15-summary`.

If we want to tag the most recent commit in the current branch
(i.e., the `HEAD`),
all we have to do is tell Git what name to give the tag:

```shell
$ git tag -a -m "It is DONE" thesis-final
```

(Please don't ever actually tempt fate by calling something `-final`.)
The `-a` option tells Git that we want to annotate the tag,
while `-m` tells it that we're providing a message on the command line.
If we want to see what tags have already been created,
we can use `git tag` on its own.
We can also tag an old commit by providing the full identifier after the tag name:

```shell
$ git tag -a -m "I should have known better" thesis-revisions 8e966b5f0703d347e4f84817f12b48f456a58ae1
```

Most software projects use [semantic versioning][semantic-versioning] for software releases,
which produces three-part version numbers `major.minor.patch`:

-   Increment `major` every time there's an incompatible externally-visible change.
-   Increment `minor` when adding functionality without breaking any existing code.
-   Increment `patch` for bug fixes that don't add any new features ("now works as previously documented").

> Simple projects only tag the `master` branch

Reviewer: ```suggestion
Simple projects should probably only tag the `master` branch
```

Author: Added "should only" (no probably)

> because everything that is finished is merged to `master`.

Reviewer: ```suggestion
because every "feature", new development, or bug fix that is finished is merged to `master`.
```

Larger software projects may create a branch for each released version and do minor or patch updates on that branch,

> but this outside the scope of this lesson.

Reviewer: ```suggestion
but this is outside the scope of this lesson.
```
